### ISSUE TITLE
Eviction callback

### ISSUE THREAD

**Issue by arossert** [dgilland/cacheout#30]
Is there any option to add an "eviction callback" when a key is deleted? 
I personally looking for a cache/TTLdict structure that I can tell when keys are evicted from it (by timeout).

**Comment by dgilland** [comment]
Unfortunately, there isn't a callback system to do that. It's something that I wanted to do but haven't been able to yet.

**Comment by blackstorm** [comment]
Think I can make a PR about the `eviction callback` function.

**Comment by arossert** [comment]
That will be awesome addition

**Comment by uncle-lv** [comment]
I'm ready to do it now. So I want to check some details for it.



This is my plan.

Declare a delete callback when create cache.

```python
cache = Cache(on_delete=on_delete)
```



`on_delete` is a callable object which is defined like this.

```python
Callable[[key: Hashable, value: Any, cause: EvictedCause],  None]
```

First parameter is the key. 

Second is the value. 

Third is the cause of cache entry eviction.



`EvictedCause` is Str or Enum which has three values:

- **EXPLICIT**:  indicates that the cache entry was explicitly deleted through an explicit operation, such as calling the cache's `delete(key: Hashable)` method.
- **EXPIRED**: indicates that the cache entry was removed because it expired.
- **SIZE**: Indicates that the cache entry was removed due to reaching the maximum size limit of the cache.



Is it OK?

**Comment by arossert** [comment]
This sounds perfect for my needs 😊

**Comment by dgilland** [comment]
Implemented in #31

**Comment by arossert** [comment]
Looks good, thanks a lot
